{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import libraries\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "import filecmp\n",
    "import subprocess\n",
    "import os\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define functions\n",
    "pattern1 = re.compile(\"GTCAG\")\n",
    "pattern_G = re.compile('[G]{2,}$')\n",
    "adapt = \"GTCAG\"\n",
    "\n",
    "def delete_Ns_before_adapter_fromR2(infile_path, outfile_path, pattern):\n",
    "    '''\n",
    "    Function removes the Ns sequence and pattern (adaptor) from the beginning of of second line of read in reads2, and move it to the header of read\n",
    "    in the form:  @header_pattern.\n",
    "    The function check length of the read and remove appropriate number of characters from fourth line of read\n",
    "    (the line with the quality of read). \n",
    "    Reads without pattern have header with \"_withoutanytail\" at the end of the end of header.\n",
    "    \n",
    "    Input:\n",
    "        infile_path: str\n",
    "        outfile_path: str\n",
    "        pattern: str\n",
    "        \n",
    "    Output: new fastq file (under output_path) withoth Ns sequence and the pattern, and truncated qualit line. \n",
    "    '''\n",
    "    with open(outfile_path, 'w') as w:\n",
    "        infile = SeqIO.parse(infile_path, \"fastq\")\n",
    "        pattern = re.compile(pattern)\n",
    "\n",
    "        for read in infile:\n",
    "            fastq = read.format(\"fastq\")\n",
    "            header = fastq.partition('\\n')[0]\n",
    "            sequence = fastq.splitlines()[1]\n",
    "            strand = fastq.splitlines()[2]\n",
    "            quality = fastq.splitlines()[3]\n",
    "            #pattern = re.compile(pattern)\n",
    "            adapter = re.search(pattern, str(sequence))\n",
    "         #   GG_seq_3prim = re.search(pattern_G, str(sequence))##\n",
    "            if adapter == None:\n",
    "                adapter1 = \"lack_of_adapter\"\n",
    "            else:\n",
    "                adapter = adapter.group(0)\n",
    "               # GG_seq_3prim = GG_seq_3prim.group(0)##\n",
    "                sequence = sequence.split(adapt, 1)[-1]\n",
    "                #sequence = \n",
    "               # NNs = sequence.split(adapt)[0]\n",
    "            \n",
    "                seq_length = len(sequence)\n",
    "                \n",
    "                quality = str(quality[-(seq_length+1) :-1])\n",
    "            #header=str(read.id) + \"_\" + NNs \n",
    "            new_read = header+'\\n'+sequence+'\\n'+strand+'\\n'+ quality+'\\n'\n",
    "            w.write(new_read)\n",
    "            \n",
    "pattern1 = re.compile(\"GTCAG\")\n",
    "pattern_G = re.compile('[G]{2,}$')\n",
    "adapt = \"GTCAG\"\n",
    "\n",
    "def move_Ns_before_adapter_fromR2_to_header(infile_path, outfile_path, pattern):\n",
    "    '''\n",
    "    Function removes the Ns sequence and pattern (adaptor) from the beginning of of second line of read in reads2, and move it to the header of read\n",
    "    in the form:  @header_pattern.\n",
    "    The function check length of the read and remove appropriate number of characters from fourth line of read\n",
    "    (the line with the quality of read). \n",
    "    Reads without pattern have header with \"_withoutanytail\" at the end of the end of header.\n",
    "    \n",
    "    Input:\n",
    "        infile_path: str\n",
    "        outfile_path: str\n",
    "        pattern: str\n",
    "        \n",
    "    Output: new fastq file (under output_path) withoth Ns sequence and the pattern, and truncated qualit line. \n",
    "    '''\n",
    "    with open(outfile_path, 'w') as w:\n",
    "        infile = SeqIO.parse(infile_path, \"fastq\")\n",
    "        pattern = re.compile(pattern)\n",
    "\n",
    "        for read in infile:\n",
    "            fastq = read.format(\"fastq\")\n",
    "            header = fastq.partition('\\n')[0]\n",
    "            sequence = fastq.splitlines()[1]\n",
    "            strand = fastq.splitlines()[2]\n",
    "            quality = fastq.splitlines()[3]\n",
    "            #pattern = re.compile(pattern)\n",
    "            adapter = re.search(pattern, str(sequence))\n",
    "            seq_raw_length = len(sequence)\n",
    "         #   GG_seq_3prim = re.search(pattern_G, str(sequence))##\n",
    "            if adapter == None:\n",
    "                adapter1 = \"lack_of_adapter\"\n",
    "                sequence_NNs = \"noNNs\"\n",
    "                header= '@'+str(read.id) + \"_\" + str(sequence_NNs)#\n",
    "                new_read = header+'\\n'+sequence+'\\n'+strand+'\\n'+ quality+'\\n'\n",
    "            else:\n",
    "                adapter = adapter.group(0)\n",
    "                sequence_NNs = sequence.split(adapt, 1)[0] \n",
    "            #GG_seq_3prim = GG_seq_3prim.group(0)##\n",
    "                sequence = sequence.split(adapt, 1)[-1]\n",
    "                \n",
    "               # NNs = sequence.split(adapt)[0]\n",
    "                seq_length = len(sequence)\n",
    "                if seq_length != seq_raw_length:\n",
    "                    quality = str(quality[-(seq_length+1) :-1])\n",
    "                else:\n",
    "                    quality = ''\n",
    "                    sequence = ''\n",
    "                header='@'+str(read.id) + \"*5NN\" + str(sequence_NNs)#\n",
    "            new_read = header+'\\n'+sequence+'\\n'+strand+'\\n'+ quality+'\\n'\n",
    "            w.write(new_read)\n",
    "            \n",
    "            \n",
    "            \n",
    "####### Function to extract tail from read sequence based on CIGAR #########\n",
    "def take_tail_fromcigar2(strand, cig, seq):\n",
    "    '''\n",
    "    Function returns tail seqence based on the CIGAR and sequence of read\n",
    "    Input:\n",
    "        cig: CIGAR from sam file after alignment qith soft-clipping  [str]\n",
    "        seq: sequence of read started from 3' end, having posibble tail   [str]\n",
    "    Output:\n",
    "        sequence of the tail extracted from read sequence based on CIGAR [str]\n",
    "    Assumptions:\n",
    "        1. For tailed reads, CIGAR starts from \\d+S (softcliping code). I's about the number\n",
    "        of softclipped nucleotides. We assume that this is also the length of the polyA tail.\n",
    "        2. Sequence seq starts from 3' tail of read.\n",
    "    '''\n",
    "\n",
    "    cigar2 = str(cig) # convert cigar to string\n",
    "    if not \"S\" in cigar2: # return nothing if CIGAR does not contain S (soft-clipping)\n",
    "        return \"\"\n",
    "    else:  # analyze data if CIGAR starts from number and S (3' soft-clipping). If not - return nothing\n",
    "        \n",
    "        \n",
    "        if strand =='-':      \n",
    "            search_result_minus = re.findall(\"[0-9]+S$\", cigar2)\n",
    "            if search_result_minus: # find CIGARs starting from 3' soft-clipping\n",
    "                number_S = int(search_result_minus[-1].split('S')[0])\n",
    "                # extract number of soft-clipped nucleotides\n",
    "                tail = seq[-number_S:] # extract tail (based of number of soft-clipped nucleotides)\n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        elif strand == '+':\n",
    "            search_result_plus = re.findall(\"^[0-9]+S\", cigar2)\n",
    "            if search_result_plus: # find CIGARs starting from 3' soft-clipping\n",
    "                number_S = int(cigar2.split('S')[0]) # extract number of soft-clipped nucleotides\n",
    "                tail = seq[0:number_S] # extract tail (based of number of soft-clipped nucleotides)\n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        else:\n",
    "             return \"\"\n",
    "        \n",
    "\n",
    "# Test the function take_tail_fromcigar():\n",
    "\n",
    "assert take_tail_fromcigar2('-', '8S', 'UUUHSAAAAAAAAAA') =='AAAAAAAA'\n",
    "assert take_tail_fromcigar2('-', '8S444M1S', 'AAAAAAAAAAUUUHS') =='S'\n",
    "assert take_tail_fromcigar2('+', '8S444M3S', 'AAAAAAAAAAUUUHS') =='AAAAAAAA'\n",
    "assert take_tail_fromcigar2('-', '67M7S', 'AAAAAAAAAAUUUHS') =='AAUUUHS'\n",
    "assert take_tail_fromcigar2('-', '*', 'AAAAAAAAAAUUUHS') ==''\n",
    "\n",
    "def def_strand(flag):\n",
    "    if flag == 4:\n",
    "        return \"unmapped\"\n",
    "    elif flag ==16:\n",
    "        return \"-\"\n",
    "    elif flag == 0:\n",
    "        return \"+\"\n",
    "    else:\n",
    "        return 'nn'\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "pattern_mRNA = re.compile(\"spac|spap|spbp|spcc|spcp|spbc\")\n",
    "pattern_tRNA = re.compile(\"trna\")\n",
    "pattern_mito = re.compile(\"spmit\")\n",
    "pattern_rRNA = re.compile(\"rrna\")\n",
    "pattern_snRNA = re.compile(\"snrna\")\n",
    "pattern_snoRNA = re.compile(\"snorna\")\n",
    "pattern_ncRNA = re.compile(\"ncrna\")\n",
    "\n",
    "\n",
    "def test_RNA_type(feature_name):\n",
    "    \"\"\"tests for the type of tail \"\"\"\n",
    "    if pattern_mRNA.search(feature_name): # for example AATT\n",
    "        return \"mRNA\"\n",
    "    elif pattern_tRNA.search(feature_name): # for example AATT\n",
    "        return \"tRNA\"\n",
    "    elif pattern_mito.search(feature_name): # for example AATT\n",
    "        return \"mito_RNA\"\n",
    "    elif pattern_rRNA.search(feature_name): # for example AATT\n",
    "        return \"rRNA\"\n",
    "    elif pattern_snRNA.search(feature_name): # for example AATT\n",
    "        return \"snRNA\"\n",
    "    elif pattern_snoRNA.search(feature_name): # for example AATT\n",
    "        return \"snoRNA\"\n",
    "    elif pattern_ncRNA.search(feature_name): # for example AATT\n",
    "        return \"ncRNA\"\n",
    "    else: #only A's\n",
    "        return \"other_type\"\n",
    "\n",
    "assert test_RNA_type(\"spncrna\") == \"ncRNA\"\n",
    "assert test_RNA_type(\"sspcc\") == \"mRNA\"\n",
    "assert test_RNA_type(\"AAAAA\") == \"other_type\"\n",
    "assert test_RNA_type(\"TAATTT\") == \"other_type\"\n",
    "\n",
    "####### Function to test tail type #########\n",
    "\n",
    "pattern_polyAU_forward = re.compile(\"^A{1,}T{2,}\") # reverse transcribed\n",
    "pattern_oligoU_forward = re.compile(\"^A{1,}\")\n",
    "pattern_polyA_forward = re.compile(\"^T{3,}$\")\n",
    "\n",
    "pattern_polyAU_reverse = re.compile(\"A{1,}T{2,}$\") # reverse transcribed\n",
    "pattern_oligoU_reverse = re.compile(\"T{1,}$\")\n",
    "pattern_polyA_reverse = re.compile(\"A{3,}$\")  \n",
    "\n",
    "def test_tail(strand, tail):\n",
    "    \"\"\" Tests for the type of tail. Patterns are prepared for analysis of R2 read, reverse transcribed:\n",
    "        base sequence start from the end of polyadenylation tail at 3' end of RNA,\n",
    "        where T means adenine, and A meand uridine in the 3' tail\"\"\"\n",
    "    if len(tail) > 0:\n",
    "        if strand == '+' or strand == 'unmapped':\n",
    "            if \"A\" in tail:\n",
    "                if pattern_polyAU_forward.fullmatch(tail): \n",
    "                    return \"polyAU\"\n",
    "                elif pattern_oligoU_forward.fullmatch(tail): \n",
    "                    return \"oligoU\"\n",
    "                else:\n",
    "                    return \"mixed_tail\"\n",
    "            elif pattern_polyA_forward.fullmatch(tail):\n",
    "                    return \"polyA\"\n",
    "            else: \n",
    "                return \"mixed_tail\"\n",
    "        elif strand == '-':\n",
    "            if \"T\" in tail:\n",
    "                if pattern_polyAU_reverse.fullmatch(tail): \n",
    "                    return \"polyAU\"\n",
    "                elif pattern_oligoU_reverse.fullmatch(tail): \n",
    "                    return \"oligoU\"\n",
    "                else:\n",
    "                    return \"mixed_tail\"\n",
    "            elif pattern_polyA_reverse.fullmatch(tail):\n",
    "                    return \"polyA\"\n",
    "            else: \n",
    "                return \"mixed_tail\"\n",
    "    else:\n",
    "        return \"without_tail\"\n",
    "    \n",
    "# Test the function test_tail():\n",
    "\n",
    "assert test_tail('+', \"AATTT\") == \"polyAU\"\n",
    "assert test_tail('+', \"AAAAAA\") == \"oligoU\"\n",
    "assert test_tail('-',\"AAAAA\") == \"polyA\"\n",
    "assert test_tail('-',\"TAATTT\") == \"mixed_tail\"\n",
    "assert test_tail('+','')== \"without_tail\"\n",
    "assert test_tail('unmapped','TTT')== \"polyA\"\n",
    "\n",
    "\n",
    "def take_tail_fromcigar2(strand, cig, seq):\n",
    "    '''\n",
    "    Function returns tail seqence based on the CIGAR and sequence of read\n",
    "    Input:\n",
    "        cig: CIGAR from sam file after alignment qith soft-clipping  [str]\n",
    "        seq: sequence of read started from 3' end, having posibble tail   [str]\n",
    "    Output:\n",
    "        sequence of the tail extracted from read sequence based on CIGAR [str]\n",
    "    Assumptions:\n",
    "        1. For tailed reads, CIGAR starts from \\d+S (softcliping code). I's about the number\n",
    "        of softclipped nucleotides. We assume that this is also the length of the polyA tail.\n",
    "        2. Sequence seq starts from 3' tail of read.\n",
    "    '''\n",
    "\n",
    "    cigar2 = str(cig) # convert cigar to string\n",
    "    if not \"S\" in cigar2: # return nothing if CIGAR does not contain S (soft-clipping)\n",
    "        return \"\"\n",
    "    else:  # analyze data if CIGAR starts from number and S (3' soft-clipping). If not - return nothing\n",
    "        \n",
    "        \n",
    "        if strand =='-':      \n",
    "            search_result_minus = re.findall(\"[0-9]+S$\", cigar2)\n",
    "            if search_result_minus: # find CIGARs starting from 3' soft-clipping\n",
    "                number_S = int(search_result_minus[-1].split('S')[0])\n",
    "                # extract number of soft-clipped nucleotides\n",
    "                tail = seq[-number_S:] # extract tail (based of number of soft-clipped nucleotides)\n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        elif strand == '+':\n",
    "            search_result_plus = re.findall(\"^[0-9]+S\", cigar2)\n",
    "            if search_result_plus: # find CIGARs starting from 3' soft-clipping\n",
    "                number_S = int(cigar2.split('S')[0]) # extract number of soft-clipped nucleotides\n",
    "                tail = seq[0:number_S] # extract tail (based of number of soft-clipped nucleotides)\n",
    "                return tail\n",
    "            else:\n",
    "                return \"\"\n",
    "        else:\n",
    "             return \"\"\n",
    "        \n",
    "\n",
    "# Test the function take_tail_fromcigar():\n",
    "\n",
    "assert take_tail_fromcigar2('-', '8S', 'UUUHSAAAAAAAAAA') =='AAAAAAAA'\n",
    "assert take_tail_fromcigar2('-', '8S444M1S', 'AAAAAAAAAAUUUHS') =='S'\n",
    "assert take_tail_fromcigar2('+', '8S444M3S', 'AAAAAAAAAAUUUHS') =='AAAAAAAA'\n",
    "assert take_tail_fromcigar2('-', '67M7S', 'AAAAAAAAAAUUUHS') =='AAUUUHS'\n",
    "assert take_tail_fromcigar2('-', '*', 'AAAAAAAAAAUUUHS') ==''\n",
    "\n",
    "\n",
    "def grep_tail_edit(strand, seq):\n",
    "    '''\n",
    "    Function ....\n",
    "    '''\n",
    "\n",
    "    if strand =='-':      \n",
    "        search_result_minus = re.findall(\"[A]{0,}[T]{0,}$\", seq)\n",
    "        if search_result_minus: # find CIGARs starting from 3' soft-clipping\n",
    "            return search_result_minus[0]\n",
    "        else:\n",
    "            return \"\"\n",
    "    elif strand == '+' or strand == 'unmapped':\n",
    "        search_result_plus = re.findall(\"^[A]{0,}[T]{0,}\", seq)\n",
    "        if search_result_plus: # find CIGARs starting from 3' soft-clipping\n",
    "            return search_result_plus[0]\n",
    "        else:\n",
    "            return \"\"\n",
    "    else:\n",
    "        return \"\"\n",
    "        \n",
    "\n",
    "# Test the function take_tail_fromcigar():\n",
    "\n",
    "assert grep_tail_edit('-', 'UUUHSAAAAAAAAAA') =='AAAAAAAAAA'\n",
    "assert grep_tail_edit('-',  'AAAAAAAAAAUUUHS') ==''\n",
    "assert grep_tail_edit('+',  'AAAAAAAAAAUUUHS') =='AAAAAAAAAA'\n",
    "\n",
    "def tail_fromGREPorCIGAR_description(cigar, tailGrep, tailCigar):\n",
    "    if cigar == '*':\n",
    "        return 'grep'\n",
    "    else:\n",
    "        return 'cigar'\n",
    "\n",
    "\n",
    "def tail_fromGREPorCIGAR(cigar, tailGrep, tailCigar):\n",
    "    if cigar == '*':\n",
    "        return tailGrep\n",
    "    else:\n",
    "        return tailCigar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analysis of Tail-seq data using both reads, R1 and R2\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of R2 reads having UMIs and adapter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/deltacid1_clone1_S77_R2_001.fastq.gz\r\n",
      "data/deltacid1_clone2_S78_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32_clone1_S85_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32_clone2_S86_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltalsm1_clone1_S97_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltalsm1_clone2_S98_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltaski2_clone1_S99_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltaski2_clone2_S100_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1_clone1_S87_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1_clone2_S88_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1deltaski2_clone1_S101_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1deltaski2_clone2_S102_R2_001.fastq.gz\r\n",
      "data/deltacid1deltaski2_clone1_S89_R2_001.fastq.gz\r\n",
      "data/deltacid1deltaski2_clone2_S90_R2_001.fastq.gz\r\n",
      "data/deltadis32_clone1_S79_R2_001.fastq.gz\r\n",
      "data/deltadis32_clone2_S80_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1_clone1_S91_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1_clone2_S92_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1deltaski2_clone1_S103_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1deltaski2_clone2_S104_R2_001.fastq.gz\r\n",
      "data/deltadis32deltaski2_clone1_S93_R2_001.fastq.gz\r\n",
      "data/deltadis32deltaski2_clone2_S94_R2_001.fastq.gz\r\n",
      "data/deltalsm1_clone1_S81_R2_001.fastq.gz\r\n",
      "data/deltalsm1_clone2_S82_R2_001.fastq.gz\r\n",
      "data/deltalsm1deltaski2_clone1_S95_R2_001.fastq.gz\r\n",
      "data/deltalsm1deltaski2_clone2_S96_R2_001.fastq.gz\r\n",
      "data/deltaski2_clone1_S83_R2_001.fastq.gz\r\n",
      "data/deltaski2_clone2_S84_R2_001.fastq.gz\r\n",
      "data/Wild_type_clone1_TS_S75_R2_001.fastq.gz\r\n",
      "data/Wild_type_clone2_TS_S76_R2_001.fastq.gz\r\n",
      "data/Wild_type_clone3_S105_R2_001.fastq.gz\r\n"
     ]
    }
   ],
   "source": [
    "!ls data/*_R2_001.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir R2_allsamples\n",
    "!mkdir R2_allsamples/filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/sh: 1: Syntax error: \"(\" unexpected\n",
      "/bin/sh: 1: Bad substitution\n"
     ]
    }
   ],
   "source": [
    "## NEWscript1_selectR2withUNIs.bash\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type\n",
    "declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Gunzip R2.fastq, select reads having 5Ns and adaptor, gzip input fastq\n",
    "for val in ${StringArray[@]}; do \n",
    "\techo \"sample_${val}\"; gunzip \"data/${val}R2_001.fastq.gz\"; grep -E -A 2 -B 1 --no-group-separator '^[A-Z][A-Z][A-Z][A-Z][A-Z][A-Z]GTCAG' \"data/${val}R2_001.fastq\" > \"R2_allsamples/filter/${val}R2_withUMIs.fastq\"; gzip \"data/${val}R2_001.fastq\"; \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/deltacid1_clone1_S77_R2_001.fastq.gz\r\n",
      "data/deltacid1_clone2_S78_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32_clone1_S85_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32_clone2_S86_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltalsm1_clone1_S97_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltalsm1_clone2_S98_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltaski2_clone1_S99_R2_001.fastq.gz\r\n",
      "data/deltacid1deltadis32deltaski2_clone2_S100_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1_clone1_S87_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1_clone2_S88_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1deltaski2_clone1_S101_R2_001.fastq.gz\r\n",
      "data/deltacid1deltalsm1deltaski2_clone2_S102_R2_001.fastq.gz\r\n",
      "data/deltacid1deltaski2_clone1_S89_R2_001.fastq.gz\r\n",
      "data/deltacid1deltaski2_clone2_S90_R2_001.fastq.gz\r\n",
      "data/deltadis32_clone1_S79_R2_001.fastq.gz\r\n",
      "data/deltadis32_clone2_S80_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1_clone1_S91_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1_clone2_S92_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1deltaski2_clone1_S103_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1deltaski2_clone2_S104_R2_001.fastq.gz\r\n",
      "data/deltadis32deltaski2_clone1_S93_R2_001.fastq.gz\r\n",
      "data/deltadis32deltaski2_clone2_S94_R2_001.fastq.gz\r\n",
      "data/deltalsm1_clone1_S81_R2_001.fastq.gz\r\n",
      "data/deltalsm1_clone2_S82_R2_001.fastq.gz\r\n",
      "data/deltalsm1deltaski2_clone1_S95_R2_001.fastq.gz\r\n",
      "data/deltalsm1deltaski2_clone2_S96_R2_001.fastq.gz\r\n",
      "data/deltaski2_clone1_S83_R2_001.fastq.gz\r\n",
      "data/deltaski2_clone2_S84_R2_001.fastq.gz\r\n",
      "data/Wild_type_clone1_TS_S75_R2_001.fastq.gz\r\n",
      "data/Wild_type_clone2_TS_S76_R2_001.fastq.gz\r\n",
      "data/Wild_type_clone3_S105_R2_001.fastq.gz\r\n"
     ]
    }
   ],
   "source": [
    "!ls data/*R2*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gunzip data/Wild_type_clone1_TS_S75_R2_001.fastq.gz\n",
    "#grep -E -A 2 -B 1 --no-group-separator '^[A-Z][A-Z][A-Z][A-Z][A-Z][A-Z]GTCAG' data/Wild_type_clone1_TS_S75_R2_001.fastq > analiza_July2022/R2/Wild_type_clone1_R2_GTCAG.fastq\n",
    "#gzip data/Wild_type_clone1_TS_S75_R2_001.fastq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " all_mapped_po1nt+kopia.fastq\r\n",
      " analiza_July2022\r\n",
      "'Analiza short.ipynb'\r\n",
      " analiza_unmapped_tailed_loop\r\n",
      " background_angeli_554genes_20220311.txt\r\n",
      " bams\r\n",
      " bams_STAR\r\n",
      " BED_R2_STAR\r\n",
      " bigwigs\r\n",
      " data\r\n",
      " deltacid1_clone1_S77_tailsinheader_20220119.fastq\r\n",
      " DESEQ_LL.R\r\n",
      " empty.fastq\r\n",
      " figures\r\n",
      " gene_cluster_20220309.csv\r\n",
      " genome\r\n",
      " genome_sdsintronsutrs\r\n",
      " gff\r\n",
      " htseq\r\n",
      " htseq_merged3.csv\r\n",
      " htseq_mRNA_doDESeq2_20220113.csv\r\n",
      " image.webp\r\n",
      " kopia_all_mapped_po1nt.fastq\r\n",
      " lista_headerow.txt\r\n",
      " Log.out\r\n",
      " mRNA_tailtypes_per_strain_20220221.csv\r\n",
      " multiqc\r\n",
      " NEWscript1_selectR2withUMIs.bash\r\n",
      " output_headertoorganise.fastq\r\n",
      " pattern_in_header.fastq\r\n",
      " polyA_pivot_dropna_20220311.csv\r\n",
      " PyPolyADetector.py\r\n",
      " R1\r\n",
      " R1b.bed\r\n",
      " R1.bed\r\n",
      " R1b.sam\r\n",
      " R1.sam\r\n",
      " R1_uniqlymappedreads_20220125.csv\r\n",
      " R1_uniqlymappedreads_procent_20220125.csv\r\n",
      " R2_allsamples\r\n",
      " R2b.bed\r\n",
      " R2b.sam\r\n",
      " R2_R1_merge_sam4_PAS.csv\r\n",
      " R2_R1_wt_test_20220712.bed\r\n",
      " R2_R1_wt_test.bed\r\n",
      " R2_R2_sam3_test_20220712.csv\r\n",
      " R2_R2_sam3_test_20220713.csv\r\n",
      " R2_R2_sam3_test_20220714.csv\r\n",
      " R2_R2_sam_test_20220712.csv\r\n",
      " R2.sam\r\n",
      " R2_uniqlymappedreads_20220125.csv\r\n",
      " R2_uniqlymappedreads_procent_20220125.csv\r\n",
      " reference\r\n",
      " robocze\r\n",
      " roboczyR1\r\n",
      " salmon\r\n",
      " sample_2.fasta\r\n",
      " sams_20220217\r\n",
      " sams_R2_STAR\r\n",
      " script10_R1_gunzipR1_cutadapt_gzip.bash\r\n",
      " script12_R1_min15nt_SSD.bash\r\n",
      " script13_STAR_frombamtounmappedfastq.bash\r\n",
      " script1_gunzipR2_select_gzip.bash\r\n",
      " script2_2_min15nt.bash\r\n",
      " script2_2_min15nt_SSD.bash\r\n",
      " script2_cutadapt2ndAdaptor.bash\r\n",
      " script3_alignmentR2_BWA.bash\r\n",
      " script3_alignmentR2_BWA_copyseqlab.bash\r\n",
      " script3_alignmentR2_STAR.bash\r\n",
      " script3b_alignmentR1_STAR.bash\r\n",
      " script3c_alignmentR2_STAR_moretails.bash\r\n",
      " script3d_77_alignmentR2_STAR_20220217.bash\r\n",
      " script3d_alignmentR2_STAR_20220217.bash\r\n",
      " script3e_alignmentR2_STAR_bezintronow_20220310.bash\r\n",
      " script4b_sam2bed_lesscolumns_STAR.bash\r\n",
      " script4_sam2bed_20220217.bash\r\n",
      " script4_sam2bed_lesscolumns.bash\r\n",
      " script4_sam2bed_ssd_20220217.bash\r\n",
      " script5_HTseq.bash\r\n",
      " script6b_bezS_closest_STAR_ssd_20220219.bash\r\n",
      " script6_closest.bash\r\n",
      " script6_closest_STAR.bash\r\n",
      " script6_closest_STAR_ssd_20220219.bash\r\n",
      " script7b_sam2bam_STAR_20220217.bash\r\n",
      " script7b_sam2bam_STAR.bash\r\n",
      " script7e_sam2bam_STAR_20220310.bash\r\n",
      " script7_sam2bam.bash\r\n",
      " script8_bam2bw_20220217.bash\r\n",
      " script8_bam2bw.bash\r\n",
      " script9_computeMatrix_plotProfile_20220217.bash\r\n",
      " script9_computeMatrix_plotProfile.bash\r\n",
      " script9_computeMatrix_plotProfile_onlymRNA.bash\r\n",
      "'Shinyplot (3).pdf'\r\n",
      " Skrypt_all_samples_LLZ_20220718.ipynb\r\n",
      " Skrypt_czysty_20220708.ipynb\r\n",
      " skrypt_July2022.ipynb\r\n",
      " sss\r\n",
      " star_data\r\n",
      " STAR_Index\r\n",
      " _STARtmp\r\n",
      " TailSeq_dec2021_manysamples_miszmasz.ipynb\r\n",
      " tails_mRNA_persample.csv\r\n",
      " test_fastq_mniej\r\n",
      " test_R1R1.csv\r\n",
      " unmapped.fastq\r\n",
      " unmapped_R2.bam\r\n",
      " unmapped_R2b.bam\r\n",
      " unmapped_R2.bed\r\n",
      " unmapped_R2b.sam\r\n",
      " unmapped_R2c.bam\r\n",
      " unmapped_R2c.sam\r\n",
      " unmapped_R2d.bam\r\n",
      " unmapped_R2d.sam\r\n",
      " unmapped_R2.sam\r\n",
      " unmapped_reads1.fastq\r\n",
      " unmapped_reads.fastq\r\n",
      "'Untitled Folder'\r\n",
      " Wild_type_clone1_tailsinheader_20220120.fastq\r\n",
      " Wild_type_clone1_tailsinheader_20220120_min15.fastq\r\n",
      " Wild_type_clone1_TS_S75_tailsinheader_20220119.fastq\r\n",
      " Wild_type_clone2_TS_S76_tailsinheader_20220119.fastq\r\n",
      " Wild_type_clone2_TS_S76_tailsinheader_20220119_min15.fastq\r\n",
      " wt_clone1_tailsinheader_20220119.fastq\r\n",
      " wt_clone2_tailsinheader_20220119.fastq\r\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(0,10):\n",
    "    os.system('ls')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove the primer GTCAG and move UMIs to the header of reads (Python script),\n",
    "#\n",
    "#move_Ns_before_adapter_fromR2_to_header(infile_path = 'analiza_July2022/R2/Wild_type_clone1_R2_GTCAG.fastq', outfile_path = 'analiza_July2022/R2/Wild_type_clone1_R2_UMIinhHeader.fastq', pattern = 'GTCAG')\n",
    "#usuwanie umi i NNs\n",
    "names = ['deltacid1_clone1_S77_',  'deltacid1_clone2_S78_',  'deltacid1deltadis32_clone1_S85_',  \n",
    "         'deltacid1deltadis32_clone2_S86_',  'deltacid1deltadis32deltalsm1_clone1_S97_',  \n",
    "         'deltacid1deltadis32deltalsm1_clone2_S98_',  'deltacid1deltadis32deltaski2_clone1_S99_',  \n",
    "         'deltacid1deltadis32deltaski2_clone2_S100_',  'deltacid1deltalsm1_clone1_S87_',  \n",
    "         'deltacid1deltalsm1_clone2_S88_',  'deltacid1deltalsm1deltaski2_clone1_S101_',  \n",
    "         'deltacid1deltalsm1deltaski2_clone2_S102_',  'deltacid1deltaski2_clone1_S89_',  \n",
    "         'deltacid1deltaski2_clone2_S90_',  'deltadis32_clone1_S79_',  'deltadis32_clone2_S80_',  \n",
    "         'deltadis32deltalsm1_clone1_S91_',  'deltadis32deltalsm1_clone2_S92_',  \n",
    "         'deltadis32deltalsm1deltaski2_clone1_S103_',  'deltadis32deltalsm1deltaski2_clone2_S104_',  \n",
    "         'deltadis32deltaski2_clone1_S93_',  'deltadis32deltaski2_clone2_S94_',  'deltalsm1_clone1_S81_',  \n",
    "         'deltalsm1_clone2_S82_',  'deltalsm1deltaski2_clone1_S95_',  'deltalsm1deltaski2_clone2_S96_', \n",
    "         'deltaski2_clone1_S83_',  'deltaski2_clone2_S84_',  'Wild_type_clone1_TS_S75_', \n",
    "         'Wild_type_clone2_TS_S76_']\n",
    "\n",
    "for name in names:\n",
    "    print(name)\n",
    "    delete_Ns_before_adapter_fromR2(infile_path = 'R2_allsamples/filter/{}R2_withUMIs.fastq'.format(name),   outfile_path = 'R2_allsamples/filter/{}R2_removedUMIs.fastq'.format(name),    \n",
    "                                    pattern = 'GTCAG')\n",
    "   # os.system('rm R2_allsamples/filter/{}R2_withUMIs.fastq'.format(name))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm R2_allsamples/filter/*_withUMIs.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of R1 reads based on R2 reads having adaptors - fast-pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rm: cannot remove 'R2_allsamples/filter/deltacid1_clone1_S77_R2_allsamples/filter/deltacid1_clone1_S77_R2_removedUMIs.fastq.paired.fq': No such file or directory\r\n",
      "rm: cannot remove 'R2_allsamples/filter/deltacid1_clone1_S77_R2_removedUMIs.fastq.single.fq': No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "###########TEST##########################\n",
    "\n",
    "## NEWscript2_selectR1fromR2_fastpair.bash\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type\n",
    "#declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Gunzip R2.fastq, select reads having 5Ns and adaptor, gzip input fastq\n",
    "!for val in 'deltacid1_clone1_S77_'; do rm \"R2_allsamples/filter/${val}R2_removedUMIs.fastq\" \"R2_allsamples/filter/${val}R2_allsamples/filter/${val}R2_removedUMIs.fastq.paired.fq\" \"R2_allsamples/filter/${val}R2_removedUMIs.fastq.single.fq\"; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## NEWscript2_selectR1fromR2_fastpair.bash\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type\n",
    "declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Gunzip R2.fastq, select reads having 5Ns and adaptor, gzip input fastq\n",
    "for val in ${StringArray[@]}; do \n",
    "\techo \"sample_${val}\"; gunzip \"data/${val}R1_001.fastq.gz\"; fastq_pair \"data/${val}R1_001.fastq\" \"R2_allsamples/filter/${val}R2_removedUMIs.fastq\"; gzip \"data/${val}R1_001.fastq\"; rm \"data/${val}R1_001.fastq.single.fq\" \"R2_allsamples/filter/${val}R2_removedUMIs.fastq\" \"R2_allsamples/filter/${val}R2_allsamples/filter/${val}R2_removedUMIs.fastq.paired.fq\" \"R2_allsamples/filter/${val}R2_removedUMIs.fastq.single.fq\"; \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gzip data/*R1_001.fastq.*.fq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##cp data/Wild_type_clone1_TS_S75_R1_001.fastq.gz analiza_July2022/R1/\n",
    "#gunzip  analiza_July2022/R1/Wild_type_clone1_TS_S75_R1_001.fastq.gz#\n",
    "#fastq_pair analiza_July2022/R1/Wild_type_clone1_TS_S75_R1_001.fastq analiza_July2022/R2/Wild_type_clone1_R2_GTCAG.fastq\n",
    "#Interesuje nas teraz plik: analiza_July2022/R1/Wild_type_clone1_TS_S75_R1_001.fastq.paired.fq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment of R1 to transcriptome using salmon\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda activate salmon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir R2_allsamples/alignmentR1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare index \n",
    "salmon index -t salmon/cds+introns+utrs.fa.gz -i salmon/pombe_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# alignment\n",
    "\n",
    "#salmon quant -i salmon/pombe_index -l MSF -r analiza_July2022/R1/Wild_type_clone1_TS_S75_R1_001.fastq.paired_cdpt2_min15.fastq.gz   -p 8 --validateMappings -o salmon/quants/Wild_type_clone1_TS_S75_quant  --writeMappings salmon/quants/Wild_type_clone1_TS_S75.sam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## NEWscript3_alignmentR1_salmon.bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type\n",
    "declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Align R1 using salmon. Generate counttable and sam file as output.\n",
    "for val in ${StringArray[@]}; do \n",
    "\techo \"sample_${val}\"; salmon quant -i salmon/pombe_index -l MSF -r  \"data/${val}_R1_001.fastq.paired.fq.gz\" -p 8 --validateMappings -o \"R2_allsamples/alignmentR1/${val}R1_quant\"  --writeMappings \"R2_allsamples/alignmentR1/${val}R1_salmon.sam\"; \n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# selection of mapped R1 reads, and converting sam to bam, and then to fastq\n",
    "#samtools view -h -o salmon/quants/Wild_type_clone1_TS_S75.bam salmon/quants/Wild_type_clone1_TS_S75.sam\n",
    "#samtools view -b -F 4 salmon/quants/Wild_type_clone1_TS_S75.bam > salmon/quants/Wild_type_clone1_TS_S75_mapped.bam\n",
    "#bedtools bamtofastq -i salmon/quants/Wild_type_clone1_TS_S75_mapped.bam -fq salmon/quants/Wild_type_clone1_TS_S75_mapped.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##NEWscript4_sam2fastq_R1.bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type\n",
    "declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Align R1 using salmon. Generate counttable and sam file as output.\n",
    "for val in ${StringArray[@]}; do \n",
    "\techo \"sample_${val}\"; samtools view -h -o \"R2_allsamples/alignmentR1/${val}R1_salmon.bam\" \"R2_allsamples/alignmentR1/${val}R1_salmon.sam\";  samtools view -b -F 4 \"R2_allsamples/alignmentR1/${val}R1_salmon.bam\" > \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.bam\";  bedtools bamtofastq -i \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.bam\" -fq \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.fastq\"; rm \"R2_allsamples/alignmentR1/${val}R1_salmon.bam\" \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.bam\"; \n",
    "done\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of reads R2, which mapped to R1 with salmon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R2_allsamples/alignmentR1/deltacid1_clone1_S77_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1_clone2_S78_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltadis32_clone1_S85_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltadis32_clone2_S86_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltadis32deltalsm1_clone1_S97_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltadis32deltalsm1_clone2_S98_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltadis32deltaski2_clone1_S99_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltadis32deltaski2_clone2_S100_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltalsm1_clone1_S87_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltalsm1_clone2_S88_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltalsm1deltaski2_clone1_S101_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltalsm1deltaski2_clone2_S102_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltaski2_clone1_S89_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltacid1deltaski2_clone2_S90_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32_clone1_S79_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32_clone2_S80_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32deltalsm1_clone1_S91_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32deltalsm1_clone2_S92_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32deltalsm1deltaski2_clone1_S103_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32deltalsm1deltaski2_clone2_S104_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32deltaski2_clone1_S93_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltadis32deltaski2_clone2_S94_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltalsm1_clone1_S81_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltalsm1_clone2_S82_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltalsm1deltaski2_clone1_S95_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltalsm1deltaski2_clone2_S96_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltaski2_clone1_S83_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/deltaski2_clone2_S84_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/Wild_type_clone1_TS_S75_R1_salmon_mapped.fastq.gz\r\n",
      "R2_allsamples/alignmentR1/Wild_type_clone2_TS_S76_R1_salmon_mapped.fastq.gz\r\n"
     ]
    }
   ],
   "source": [
    "!ls R2_allsamples/alignmentR1/*R1_salmon_mapped.fastq*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######NEWscript5_fastpair_mappedR1_R2.bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type\n",
    "declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Align R1 using salmon. Generate counttable and sam file as output.\n",
    "for val in ${StringArray[@]}; do \n",
    "\techo \"sample_${val}\"; gunzip \"data/${val}R2_001.fastq.gz\" \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.fastq\"; fastq_pair \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.fastq\" \"data/${val}R2_001.fastq\"; gzip \"data/${val}R2_001.fastq.gz\" \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.fastq\"; rm \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.fastq.paired.fq\" \"R2_allsamples/alignmentR1/${val}R1_salmon_mapped.fastq.single.fq\" \"data/${val}R2_001.fastq.single.fq\"; \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fastq_pair salmon/quants/Wild_type_clone1_TS_S75_mapped.fastq analiza_July2022/R2/Wild_type_clone1_R2_GTCAG.fastq\n",
    "#important output: analiza_July2022/R2/Wild_type_clone1_R2_GTCAG.fastq.paired.fq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment of selected R2 with STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir R2_allsamples/alignmentR2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "deltacid1deltalsm1_clone1_S87_\n",
      "deltacid1deltalsm1_clone2_S88_\n",
      "deltacid1deltalsm1deltaski2_clone1_S101_\n",
      "deltacid1deltalsm1deltaski2_clone2_S102_\n",
      "deltacid1deltaski2_clone1_S89_\n",
      "deltacid1deltaski2_clone2_S90_\n",
      "deltadis32_clone1_S79_\n",
      "deltadis32_clone2_S80_\n",
      "deltadis32deltalsm1_clone1_S91_\n",
      "deltadis32deltalsm1_clone2_S92_\n",
      "deltadis32deltalsm1deltaski2_clone1_S103_\n",
      "deltadis32deltalsm1deltaski2_clone2_S104_\n",
      "deltadis32deltaski2_clone1_S93_\n",
      "deltadis32deltaski2_clone2_S94_\n",
      "deltalsm1_clone1_S81_\n",
      "deltalsm1_clone2_S82_\n",
      "deltalsm1deltaski2_clone1_S95_\n",
      "deltalsm1deltaski2_clone2_S96_\n",
      "deltaski2_clone1_S83_\n",
      "deltaski2_clone2_S84_\n",
      "Wild_type_clone1_TS_S75_\n",
      "Wild_type_clone2_TS_S76_\n"
     ]
    }
   ],
   "source": [
    "# preparation of data \n",
    "names = ['deltacid1_clone1_S77_',  'deltacid1_clone2_S78_',  'deltacid1deltadis32_clone1_S85_',  \n",
    "         'deltacid1deltadis32_clone2_S86_',  'deltacid1deltadis32deltalsm1_clone1_S97_',  \n",
    "         'deltacid1deltadis32deltalsm1_clone2_S98_',  'deltacid1deltadis32deltaski2_clone1_S99_',  \n",
    "         'deltacid1deltadis32deltaski2_clone2_S100_',  'deltacid1deltalsm1_clone1_S87_',  \n",
    "         'deltacid1deltalsm1_clone2_S88_',  'deltacid1deltalsm1deltaski2_clone1_S101_',  \n",
    "         'deltacid1deltalsm1deltaski2_clone2_S102_',  'deltacid1deltaski2_clone1_S89_',  \n",
    "         'deltacid1deltaski2_clone2_S90_',  'deltadis32_clone1_S79_',  'deltadis32_clone2_S80_',  \n",
    "         'deltadis32deltalsm1_clone1_S91_',  'deltadis32deltalsm1_clone2_S92_',  \n",
    "         'deltadis32deltalsm1deltaski2_clone1_S103_',  'deltadis32deltalsm1deltaski2_clone2_S104_',  \n",
    "         'deltadis32deltaski2_clone1_S93_',  'deltadis32deltaski2_clone2_S94_',  'deltalsm1_clone1_S81_',  \n",
    "         'deltalsm1_clone2_S82_',  'deltalsm1deltaski2_clone1_S95_',  'deltalsm1deltaski2_clone2_S96_', \n",
    "         'deltaski2_clone1_S83_',  'deltaski2_clone2_S84_',  'Wild_type_clone1_TS_S75_', \n",
    "         'Wild_type_clone2_TS_S76_']\n",
    "\n",
    "for name in names:\n",
    "    print(name)\n",
    "    move_Ns_before_adapter_fromR2_to_header(infile_path = 'data/{}R2_001.fastq.paired.fq'.format(name),\n",
    "                                         outfile_path = 'R2_allsamples/alignmentR2/{}R2_pairedtoR1.fastq'.format(name),\n",
    "                                            pattern = 'GTCAG')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## NEWscript6_alignment_R2_STAR.bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type#\n",
    "declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Align R1 using salmon. Generate counttable and sam file as output.\n",
    "for val in ${StringArray[@]}; do \n",
    "\techo \"sample_${val}\"; STAR --genomeDir genome/ --readFilesIn \"R2_allsamples/alignmentR2/${val}R2_pairedtoR1.fastq\" --alignIntronMax 1000 --outSAMtype BAM Unsorted --outFileNamePrefix \"R2_allsamples/alignmentR2/${val}R2_STAR_\" –outFilterMultimapNmax 1--outSAMunmapped Within; gzip \"R2_allsamples/alignmentR2/${val}R2_pairedtoR1.fastq\"; \n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Down stream data analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!samtools view -h -o analiza_July2022/R2/Wild_type_clone1_R2_tylkozmapzR1_UMIinheader_Aligned.out.sam analiza_July2022/R2/Wild_type_clone1_R2_tylkozmapzR1_UMIinheader_Aligned.out.bam\n",
    "\n",
    "#sam to bed\n",
    "!#sam2bed < analiza_July2022/R2/Wild_type_clone1_R2_tylkozmapzR1_UMIinheader_Aligned.out.sam > analiza_July2022/R2/Wild_type_clone1_R2_tylkozmapzR1_UMIinheader_Aligned.out.bed\n",
    "## wersja z bed zamiast z sam\n",
    "#!awk -F\"\\t\" '{print $1,$2,$3,$4,$6,$8,$9,$12}' analiza_July2022/R2/Wild_type_clone1_R2_tylkozmapzR1_UMIinheader_Aligned.out.bed > R2b.bed\n",
    "\n",
    "### NEWscript7\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# Declare an array of string with type#\n",
    "declare -a StringArray=('deltacid1_clone1_S77_' 'deltacid1_clone2_S78_' 'deltacid1deltadis32_clone1_S85_' 'deltacid1deltadis32_clone2_S86_' 'deltacid1deltadis32deltalsm1_clone1_S97_' 'deltacid1deltadis32deltalsm1_clone2_S98_' 'deltacid1deltadis32deltaski2_clone1_S99_' 'deltacid1deltadis32deltaski2_clone2_S100_' 'deltacid1deltalsm1_clone1_S87_' 'deltacid1deltalsm1_clone2_S88_' 'deltacid1deltalsm1deltaski2_clone1_S101_' 'deltacid1deltalsm1deltaski2_clone2_S102_' 'deltacid1deltaski2_clone1_S89_' 'deltacid1deltaski2_clone2_S90_' 'deltadis32_clone1_S79_' 'deltadis32_clone2_S80_' 'deltadis32deltalsm1_clone1_S91_' 'deltadis32deltalsm1_clone2_S92_' 'deltadis32deltalsm1deltaski2_clone1_S103_' 'deltadis32deltalsm1deltaski2_clone2_S104_' 'deltadis32deltaski2_clone1_S93_' 'deltadis32deltaski2_clone2_S94_' 'deltalsm1_clone1_S81_' 'deltalsm1_clone2_S82_' 'deltalsm1deltaski2_clone1_S95_' 'deltalsm1deltaski2_clone2_S96_' 'deltaski2_clone1_S83_' 'deltaski2_clone2_S84_' 'Wild_type_clone1_TS_S75_' 'Wild_type_clone2_TS_S76_')\n",
    "\n",
    "# Iterate the string array using for loop\n",
    "# Align R1 using salmon. Generate counttable and sam file as output.\n",
    "for val in ${StringArray[@]}; do \n",
    "\techo \"sample_${val}\"; samtools view -h -o \"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.sam\" \"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.bam\"; gzip \"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.bam\"; sam2bed < \"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.sam\" >\"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.bed\"; gzip \"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.sam\"; awk -F\"\\t\" '{print $1,$2,$3,$4,$6,$8,$9,$12}' \"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.bed\" > \"R2_allsamples/alignmentR2/${val}R2b.bed\"; gzip \"R2_allsamples/alignmentR2/${val}R2_STAR_Aligned.out.bed\"; \n",
    "done\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AB325691 1306 1381 A00805:117:HKV7NDRXY:2:2107:23430:28385*5NNGGGCGG + 15S75M * TTTTTTTTTTTTTTTTAGATGGTCAGAAAGTTTTTCGGGAAGATATCGTAAGTATATCAAGTTTGNTGTGACACGTGTAAACGGCAAACA\r\n",
      "AB325691 2063 2144 A00805:117:HKV7NDRXY:1:1125:16161:9126*5NNGAGTAC + 81M9S * ACCAGCCATACCGTGAAGGGTACGAGCGGCTGCGTAAGTGCTAGAATTAGTTACGGAAACGACAGATATAATAATTACAGCTGTCTCTTA\r\n",
      "AB325691 2901 2991 A00805:117:HKV7NDRXY:2:1229:15094:9502*5NNTAGTGG - 90M * CCATCTGAAAGTGAACTGCCTGAACCAACGTAAAGACCACTGCCGATAGCCCCCCCAATTGCGATCATTTGCATATGTCGAGCTTTTAAG\r\n",
      "AB325691 8790 8860 A00805:117:HKV7NDRXY:1:2269:28691:29105*5NNC + 23S70M2S * GTCAGTTTTTTTTTTTTTTTTTTTTTGTGAGAGAACATTTTATTAGAGTTATATAAAACATTATCTGGAGCATAAACCCTCAATCTCTTTAATAA\r\n",
      "AB325691 8790 8864 A00805:117:HKV7NDRXY:1:2251:5177:19914*5NNCTTGTT + 16S74M * TTTTTTTTTTTTTTTTTTTTTGAGAGAACATTTTATTAGATTTATATAAAACATTAGCTGGAGCATAAAACCTCAATATATTTAATTATT\r\n",
      "AB325691 8790 8864 A00805:117:HKV7NDRXY:1:2251:5529:19492*5NNCTTGTT + 16S74M * TTTTTTTTTTTTTTTTTTTTTGAGAGAACATTTTATTATAGTTATATAAAAAAGTATCTTGAGCATAAACCCTCAATATCTTTAATTATT\r\n",
      "AB325691 8790 8865 A00805:117:HKV7NDRXY:1:2163:13331:14246*5NNCTTCTT + 14S75M1S * TTTTTTTTTTTTTTTTTGTGAGAGAACATTTTATTATAGTTATATAAAACAGTAGCTGGAGCATAAACCCTCAATCTCTTTAATTATTAT\r\n",
      "AB325691 8790 8865 A00805:117:HKV7NDRXY:2:2143:16550:28745*5NNC + 19S75M1S * GTCAGTTTTTTTTTTTTTTTTTGTGAGAGAACATTTTATTAGAGTTATATAAAACAGTAGCTGGAGCATAAACCCTCAATCACTTTAATTATTAT\r\n",
      "AB325691 8790 8866 A00805:117:HKV7NDRXY:1:1275:17933:3020*5NNGTTCGC + 14S76M * TTTTTTTTTTTTTTTTTGTGAGAGAACATTTTATTAGATTTATATAAAACAGTAGCTGGAGCATAAACCCTCAATCTCTTTAATTATTAG\r\n",
      "AB325691 8790 8866 A00805:117:HKV7NDRXY:1:1275:18719:5040*5NNGTTCGC + 14S76M * TTTTTTTTTTTTTTTTTGTGAGAGAACATTTTATTAGAGTTATATAAAACAGTAGCTGGAGCATAAACCCTCAATCTCTTTAATTATTAG\r\n"
     ]
    }
   ],
   "source": [
    "!head R2_allsamples/alignmentR2/deltacid1_clone1_S77_R2b.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## R1\n",
    "#!grep A00805 salmon/quants/Wild_type_clone1_TS_S75.sam > R1.sam\n",
    "#!sam2bed < R1.sam > R1.bed\n",
    "#!awk -F\"\\t\" '{print $1,$4,$6,$7}' R1.bed > R1b.bed\n",
    "#!awk -F\"\\t\" '{print $1,$2,$3}' R1.sam > R1b.sam\n",
    "\n",
    "### NEWscript8\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12799 reference/newSP_gene.bed\r\n"
     ]
    }
   ],
   "source": [
    "## preparation reference table with gene names and coordinates\n",
    "!grep gene reference/newSP.bed > reference/newSP_gene.bed\n",
    "!wc -l reference/newSP_gene.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>5662</td>\n",
       "      <td>SPAC212.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>345</td>\n",
       "      <td>1164</td>\n",
       "      <td>SPNCRNA.2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>2117</td>\n",
       "      <td>2347</td>\n",
       "      <td>SPNCRNA.2001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>2447</td>\n",
       "      <td>2760</td>\n",
       "      <td>SPNCRNA.2002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>4056</td>\n",
       "      <td>4311</td>\n",
       "      <td>SPNCRNA.2003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr  start  stop          gene\n",
       "0   I      0  5662    SPAC212.11\n",
       "1   I    345  1164  SPNCRNA.2000\n",
       "2   I   2117  2347  SPNCRNA.2001\n",
       "3   I   2447  2760  SPNCRNA.2002\n",
       "4   I   4056  4311  SPNCRNA.2003"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## preparation reference table with gene names and coordinates\n",
    "\n",
    "reference_newSP = pd.read_csv('reference/newSP_gene.bed', names=['chr', 'start', 'stop', 'gene','1','2','3',\n",
    "                                                           '4', '5', '6'], sep = '\\t')\n",
    "\n",
    "reference_newSP = reference_newSP.drop(columns=['1','2','3','4', '5', '6'])\n",
    "reference_newSP.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gzip R2_allsamples/alignmentR2/deltacid1_clone1_S77_R2_STAR_Aligned.out.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir R2_allsamples/dfs_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "deltadis32deltaski2_clone1_S93_\n",
      "deltadis32deltaski2_clone2_S94_\n",
      "deltalsm1_clone1_S81_\n",
      "deltalsm1_clone2_S82_\n",
      "deltalsm1deltaski2_clone1_S95_\n",
      "deltalsm1deltaski2_clone2_S96_\n",
      "deltaski2_clone1_S83_\n",
      "deltaski2_clone2_S84_\n",
      "Wild_type_clone1_TS_S75_\n",
      "Wild_type_clone2_TS_S76_\n"
     ]
    }
   ],
   "source": [
    "# preparation of data \n",
    "names = [#'deltacid1_clone1_S77_',  'deltacid1_clone2_S78_',  'deltacid1deltadis32_clone1_S85_',  \n",
    "         #'deltacid1deltadis32_clone2_S86_', 'deltacid1deltadis32deltalsm1_clone1_S97_',  \n",
    "         #'deltacid1deltadis32deltalsm1_clone2_S98_',\n",
    "   # 'deltacid1deltadis32deltaski2_clone1_S99_',  \n",
    "      #   'deltacid1deltadis32deltaski2_clone2_S100_',  'deltacid1deltalsm1_clone1_S87_',  \n",
    "        # 'deltacid1deltalsm1_clone2_S88_',  'deltacid1deltalsm1deltaski2_clone1_S101_',  \n",
    "        # 'deltacid1deltalsm1deltaski2_clone2_S102_',  'deltacid1deltaski2_clone1_S89_',  \n",
    "        # 'deltacid1deltaski2_clone2_S90_',  'deltadis32_clone1_S79_',  'deltadis32_clone2_S80_',  \n",
    "       #  'deltadis32deltalsm1_clone1_S91_', \n",
    "         \n",
    "       #  'deltadis32deltalsm1_clone2_S92_',  \n",
    "       #  'deltadis32deltalsm1deltaski2_clone1_S103_',\n",
    "     ######### $$$$$$$$$$ 'deltadis32deltalsm1deltaski2_clone2_S104_',\n",
    "    \n",
    "         'deltadis32deltaski2_clone1_S93_',  'deltadis32deltaski2_clone2_S94_',  'deltalsm1_clone1_S81_',  \n",
    "         'deltalsm1_clone2_S82_',  'deltalsm1deltaski2_clone1_S95_',  'deltalsm1deltaski2_clone2_S96_', \n",
    "         'deltaski2_clone1_S83_',  'deltaski2_clone2_S84_',  'Wild_type_clone1_TS_S75_', \n",
    "         'Wild_type_clone2_TS_S76_'\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "for name in names:\n",
    "    print(name)\n",
    "    r2_sam = pd.read_csv('R2_allsamples/alignmentR2/{}R2_shortsam.sam'.format(name),\n",
    "                         sep = ' ',names=['readID', 'flag', 'chr','coord', 'num', 'cigar', 'seq'])\n",
    "    r2_sam = r2_sam[(r2_sam['flag']<200)]\n",
    "    r2_sam['strand'] = r2_sam['flag'].apply(lambda x: def_strand(x))\n",
    "    r2_sam['UMIs'] = r2_sam['readID'].apply(lambda x: x.split('*5NN')[-1])\n",
    "    r2_sam['UMIs_seq'] = r2_sam['UMIs'] + '_' + r2_sam['seq']\n",
    "\n",
    "    r2_sam['readID'] = r2_sam['readID'].apply(lambda x: x.split('*5NN')[0])\n",
    "    r2_sam  = r2_sam.drop(columns = ['flag'])\n",
    "    #r2_sam.to_csv('R2_allsamples/alignmentR2/{}R2_sam.csv'.format(name)\n",
    "    \n",
    "    r2_bed = pd.read_csv('R2_allsamples/alignmentR2/{}R2b.bed'.format(name),\n",
    "                         sep = ' ',names=['chr','start', 'stop','readID', 'strand','cigar','*', 'seq'])\n",
    "    r2_bed['start_read'] = r2_bed['start']\n",
    "    r2_bed['stop_read'] = r2_bed['stop']\n",
    "    r2_bed['readID'] = r2_bed['readID'].apply(lambda x: x.split('*5NN')[0])\n",
    "    r2_bed = r2_bed.drop(columns=['chr', 'start', 'stop','strand', 'cigar', '*', 'seq'])\n",
    "    r2_bed = r2_bed.drop_duplicates()\n",
    "    \n",
    "    r1_bed = pd.read_csv('R2_allsamples/alignmentR1/{}R1b.bed'.format(name), sep = ' ',names=['gene','readID','strand','flag_R1'])\n",
    "    r1_bed = r1_bed[r1_bed['flag_R1']<2000]\n",
    "    r1_bed.drop_duplicates()\n",
    "    #r2_sam.drop_duplicates()\n",
    "    r1_sam = pd.read_csv(\"R2_allsamples/alignmentR1/{}R1b.sam\".format(name), sep = ' ',names=['readID','flag_R1', 'gene_R1'])\n",
    "    r1_sam_bezzbednychflag = r1_sam[r1_sam['flag_R1']<2000]\n",
    "    r3 = r2_sam.drop(columns=[  'num'])\n",
    "    r3 = r3.drop_duplicates()\n",
    "\n",
    "    R2_R1_merge_sam = pd.merge(r3, r1_bed, on = ['readID'])\n",
    "#R2_R1_merge_sam2 = pd.merge(R2_R1_merge_sam, r2_bed, on = ['readID'])\n",
    "    R2_R1_merge_sam2 = R2_R1_merge_sam.drop_duplicates()\n",
    "    R2_R1_merge_sam2.head()\n",
    "\n",
    "    R2_R1_merge_sam3_in = R2_R1_merge_sam2.set_index('readID')\n",
    "    r2_bed_in = r2_bed.set_index('readID')\n",
    "\n",
    "    R2_R1_merge_sam4 = R2_R1_merge_sam3_in.join(r2_bed_in)\n",
    "    R2_R1_merge_sam4.to_csv('R2_allsamples/dfs_output/df_{}.csv'.format(name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rwxrwxrwx 1 lidia lidia 538M lip 21 11:46 R2_allsamples/dfs_output/df_Wild_type_clone2_TS_S76_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 513M lip 21 11:45 R2_allsamples/dfs_output/df_Wild_type_clone1_TS_S75_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 586M lip 21 11:44 R2_allsamples/dfs_output/df_deltaski2_clone2_S84_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 642M lip 21 11:43 R2_allsamples/dfs_output/df_deltaski2_clone1_S83_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 913M lip 21 11:41 R2_allsamples/dfs_output/df_deltalsm1deltaski2_clone2_S96_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 636M lip 21 11:39 R2_allsamples/dfs_output/df_deltalsm1deltaski2_clone1_S95_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 850M lip 21 11:38 R2_allsamples/dfs_output/df_deltalsm1_clone2_S82_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 804M lip 21 11:36 R2_allsamples/dfs_output/df_deltalsm1_clone1_S81_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 862M lip 21 11:35 R2_allsamples/dfs_output/df_deltadis32deltaski2_clone2_S94_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 1,2G lip 21 11:34 R2_allsamples/dfs_output/df_deltadis32deltaski2_clone1_S93_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 674M lip 21 11:21 R2_allsamples/dfs_output/df_deltadis32deltalsm1deltaski2_clone1_S103_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 1,6G lip 21 11:20 R2_allsamples/dfs_output/df_deltadis32deltalsm1_clone2_S92_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 644M lip 21 11:18 R2_allsamples/dfs_output/df_deltadis32deltalsm1_clone1_S91_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 692M lip 21 11:16 R2_allsamples/dfs_output/df_deltadis32_clone2_S80_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 557M lip 21 11:14 R2_allsamples/dfs_output/df_deltadis32_clone1_S79_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 570M lip 21 11:13 R2_allsamples/dfs_output/df_deltacid1deltaski2_clone2_S90_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 484M lip 21 11:10 R2_allsamples/dfs_output/df_deltacid1deltaski2_clone1_S89_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 585M lip 21 11:07 R2_allsamples/dfs_output/df_deltacid1deltalsm1deltaski2_clone2_S102_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 1,3G lip 21 11:05 R2_allsamples/dfs_output/df_deltacid1deltalsm1deltaski2_clone1_S101_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 1,1G lip 21 11:00 R2_allsamples/dfs_output/df_deltacid1deltalsm1_clone2_S88_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 794M lip 21 10:56 R2_allsamples/dfs_output/df_deltacid1deltalsm1_clone1_S87_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 712M lip 21 10:53 R2_allsamples/dfs_output/df_deltacid1deltadis32deltaski2_clone2_S100_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 1,1G lip 21 10:51 R2_allsamples/dfs_output/df_deltacid1deltadis32deltaski2_clone1_S99_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 906M lip 21 10:44 R2_allsamples/dfs_output/df_deltacid1deltadis32deltalsm1_clone2_S98_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 1,1G lip 21 10:41 R2_allsamples/dfs_output/df_deltacid1deltadis32deltalsm1_clone1_S97_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 702M lip 21 10:37 R2_allsamples/dfs_output/df_deltacid1deltadis32_clone2_S86_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 671M lip 21 10:34 R2_allsamples/dfs_output/df_deltacid1deltadis32_clone1_S85_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 550M lip 21 10:32 R2_allsamples/dfs_output/df_deltacid1_clone2_S78_.csv\r\n",
      "-rwxrwxrwx 1 lidia lidia 532M lip 21 10:31 R2_allsamples/dfs_output/df_deltacid1_clone1_S77_.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls -lth R2_allsamples/dfs_output/df*.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readID,chr,coord,cigar,seq,strand_x,UMIs,UMIs_seq,gene,strand_y,flag_R1,start_read,stop_read\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10031:31328,I,3344284,90M,TATACAATTTTTTGAATCCAAAAACAGTTACAGAATATTTCAAGAATAGTGTTTAATATATTCAAATGCTTATTTTTAACAAAAAAAAAA,-,ACCGAC,ACCGAC_TATACAATTTTTTGAATCCAAAAACAGTTACAGAATATTTCAAGAATAGTGTTTAATATATTCAAATGCTTATTTTTAACAAAAAAAAAA,SPAC26A3.05,+,0,3344283,3344373\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10086:3145,II,1513300,90M,CTGATTTTACAAGCAACATTACTAACAATTACAGGCCGCGAATATATTACTTAGCTCAGTTTAGTGTATGTGAGAATCAATCAATTATGA,+,CACCAC,CACCAC_CTGATTTTACAAGCAACATTACTAACAATTACAGGCCGCGAATATATTACTTAGCTCAGTTTAGTGTATGTGAGAATCAATCAATTATGA,SPBC83.01,-,16,1513299,1513389\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10122:5838,I,4268827,90M,ATGCACATCACTACCGGGTCTTGGGCAGTGCGATAGCGATGGGATTCACCTTCGCAGGATGTGCATGGAAGTATAAACACAACGGTCGTT,-,TCTGAC,TCTGAC_ATGCACATCACTACCGGGTCTTGGGCAGTGCGATAGCGATGGGATTCACCTTCGCAGGATGTGCATGGAAGTATAAACACAACGGTCGTT,SPNCRNA.3909,-,16,4268826,4268916\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10131:11741,II,3958833,90M,ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGAACATATCATGTAAGAGATCATCGCAAATAAAGATAGCACAACAA,+,GCTAAG,GCTAAG_ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGAACATATCATGTAAGAGATCATCGCAAATAAAGATAGCACAACAA,SPSNRNA.07,+,0,3958832,3958922\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10131:20635,II,3958833,80M10S,ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGAACATATCATGTAAGAGATCATCGCAAATAAAGATCTGTCTCTTA,+,TGGGTC,TGGGTC_ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGAACATATCATGTAAGAGATCATCGCAAATAAAGATCTGTCTCTTA,SPSNRNA.07,+,0,3958832,3958912\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10158:28009,III,16914,90M,ACAAATCTTGGGAACAAAGGCTTAATCTCAGCAGATCGTAACAACAAGGCTACTCTACTGCTTACAATACCCCGTTCCACATTAAAGTCG,+,ACCCCA,ACCCCA_ACAAATCTTGGGAACAAAGGCTTAATCTCAGCAGATCGTAACAACAAGGCTACTCTACTGCTTACAATACCCCGTTCCACATTAAAGTCG,SPNCRNA.6669,-,16,6043,6133\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10158:28009,III,16914,90M,ACAAATCTTGGGAACAAAGGCTTAATCTCAGCAGATCGTAACAACAAGGCTACTCTACTGCTTACAATACCCCGTTCCACATTAAAGTCG,+,ACCCCA,ACCCCA_ACAAATCTTGGGAACAAAGGCTTAATCTCAGCAGATCGTAACAACAAGGCTACTCTACTGCTTACAATACCCCGTTCCACATTAAAGTCG,SPNCRNA.6669,-,16,16913,17003\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10158:28009,III,16914,90M,ACAAATCTTGGGAACAAAGGCTTAATCTCAGCAGATCGTAACAACAAGGCTACTCTACTGCTTACAATACCCCGTTCCACATTAAAGTCG,+,ACCCCA,ACCCCA_ACAAATCTTGGGAACAAAGGCTTAATCTCAGCAGATCGTAACAACAAGGCTACTCTACTGCTTACAATACCCCGTTCCACATTAAAGTCG,SPNCRNA.6669,-,16,2447120,2447210\r\n",
      "A00805:117:HKV7NDRXY:1:1101:10203:18912,I,1234763,88M2S,TGAATTGTGGTTTTTCGTCGTAGTTGATAGTTATACGGTCGCATCCATTTGTTGATGTATCACAATGGGGCTTAGTCTCGTGCTCAAAAA,-,GCAGCC,GCAGCC_TGAATTGTGGTTTTTCGTCGTAGTTGATAGTTATACGGTCGCATCCATTTGTTGATGTATCACAATGGGGCTTAGTCTCGTGCTCAAAAA,SPNCRNA.2541,-,16,1234762,1234850\r\n"
     ]
    }
   ],
   "source": [
    "!head R2_allsamples/dfs_output/df_*.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "r2_bed = pd.read_csv('R2b.bed', sep = ' ',names=['chr','start', 'stop','readID', 'strand','cigar','*', 'seq'])\n",
    "r2_bed['start_read'] = r2_bed['start']\n",
    "r2_bed['stop_read'] = r2_bed['stop']\n",
    "r2_bed['readID'] = r2_bed['readID'].apply(lambda x: x.split('*5NN')[0])\n",
    "r2_bed = r2_bed.drop(columns=['chr', 'start', 'stop','strand', 'cigar', '*', 'seq'])\n",
    "r2_bed = r2_bed.drop_duplicates()\n",
    "\n",
    "\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "r1_bed = pd.read_csv('R1b.bed', sep = ' ',names=['gene','readID','strand','flag_R1'])\n",
    "r1_bed = r1_bed[r1_bed['flag_R1']<2000]\n",
    "r1_bed.drop_duplicates()\n",
    "\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "!head reference/newSP.bed\n",
    "reference_newSP = pd.read_csv('reference/newSP.bed', names=['chr', 'start', 'stop', 'gene','1','2','3',\n",
    "                                                           '4', '5', '6'], sep = '\\t')\n",
    "\n",
    "reference_newSP = reference_newSP.drop(columns=['1','2','3','4', '5', '6'])\n",
    "reference_newSP.head()\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#r1_sam = pd.read_csv('R1b.sam', sep = ' ',names=['readID','flag_R1', 'gene_R1'])\n",
    "#r1_sam_bezzbednychflag = r1_sam[r1_sam['flag_R1']<2000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#r3 = r2_sam.drop(columns=[  'num'])\n",
    "#r3 = r3.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#R2_R1_merge_sam = pd.merge(r3, r1_bed, on = ['readID'])\n",
    "#R2_R1_merge_sam2 = pd.merge(R2_R1_merge_sam, r2_bed, on = ['readID'])\n",
    "#R2_R1_merge_sam2 = R2_R1_merge_sam.drop_duplicates()\n",
    "#R2_R1_merge_sam2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readID</th>\n",
       "      <th>chr</th>\n",
       "      <th>coord</th>\n",
       "      <th>seq</th>\n",
       "      <th>strand_x</th>\n",
       "      <th>UMIs</th>\n",
       "      <th>UMIs_seq</th>\n",
       "      <th>gene</th>\n",
       "      <th>strand_y</th>\n",
       "      <th>flag_R1</th>\n",
       "      <th>start_read</th>\n",
       "      <th>stop_read</th>\n",
       "      <th>RNA_type</th>\n",
       "      <th>tail_fromcigar</th>\n",
       "      <th>tail_fromGREP</th>\n",
       "      <th>tail_GreporCigar</th>\n",
       "      <th>tail_from</th>\n",
       "      <th>tail_type_FINAL</th>\n",
       "      <th>tail_len_FINAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00805:117:HKV7NDRXY:1:1101:10031:31328</td>\n",
       "      <td>I</td>\n",
       "      <td>3344284</td>\n",
       "      <td>TATACAATTTTTTGAATCCAAAAACAGTTACAGAATATTTCAAGAA...</td>\n",
       "      <td>-</td>\n",
       "      <td>ACCGAC</td>\n",
       "      <td>ACCGAC_TATACAATTTTTTGAATCCAAAAACAGTTACAGAATATT...</td>\n",
       "      <td>SPAC26A3.05</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>3344283</td>\n",
       "      <td>3344373</td>\n",
       "      <td>mRNA</td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>without_tail</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00805:117:HKV7NDRXY:1:1101:10086:3145</td>\n",
       "      <td>II</td>\n",
       "      <td>1513300</td>\n",
       "      <td>CTGATTTTACAAGCAACATTACTAACAATTACAGGCCGCGAATATA...</td>\n",
       "      <td>+</td>\n",
       "      <td>CACCAC</td>\n",
       "      <td>CACCAC_CTGATTTTACAAGCAACATTACTAACAATTACAGGCCGC...</td>\n",
       "      <td>SPBC83.01</td>\n",
       "      <td>-</td>\n",
       "      <td>16</td>\n",
       "      <td>1513299</td>\n",
       "      <td>1513389</td>\n",
       "      <td>mRNA</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>without_tail</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00805:117:HKV7NDRXY:1:1101:10122:5838</td>\n",
       "      <td>I</td>\n",
       "      <td>4268827</td>\n",
       "      <td>ATGCACATCACTACCGGGTCTTGGGCAGTGCGATAGCGATGGGATT...</td>\n",
       "      <td>-</td>\n",
       "      <td>TCTGAC</td>\n",
       "      <td>TCTGAC_ATGCACATCACTACCGGGTCTTGGGCAGTGCGATAGCGA...</td>\n",
       "      <td>SPNCRNA.3909</td>\n",
       "      <td>-</td>\n",
       "      <td>16</td>\n",
       "      <td>4268826</td>\n",
       "      <td>4268916</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td></td>\n",
       "      <td>TT</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>without_tail</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00805:117:HKV7NDRXY:1:1101:10131:11741</td>\n",
       "      <td>II</td>\n",
       "      <td>3958833</td>\n",
       "      <td>ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGA...</td>\n",
       "      <td>+</td>\n",
       "      <td>GCTAAG</td>\n",
       "      <td>GCTAAG_ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACC...</td>\n",
       "      <td>SPSNRNA.07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>3958832</td>\n",
       "      <td>3958922</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>without_tail</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A00805:117:HKV7NDRXY:1:1101:10131:20635</td>\n",
       "      <td>II</td>\n",
       "      <td>3958833</td>\n",
       "      <td>ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGA...</td>\n",
       "      <td>+</td>\n",
       "      <td>TGGGTC</td>\n",
       "      <td>TGGGTC_ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACC...</td>\n",
       "      <td>SPSNRNA.07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>3958832</td>\n",
       "      <td>3958912</td>\n",
       "      <td>snRNA</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td></td>\n",
       "      <td>cigar</td>\n",
       "      <td>without_tail</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    readID chr    coord  \\\n",
       "0  A00805:117:HKV7NDRXY:1:1101:10031:31328   I  3344284   \n",
       "1   A00805:117:HKV7NDRXY:1:1101:10086:3145  II  1513300   \n",
       "2   A00805:117:HKV7NDRXY:1:1101:10122:5838   I  4268827   \n",
       "3  A00805:117:HKV7NDRXY:1:1101:10131:11741  II  3958833   \n",
       "4  A00805:117:HKV7NDRXY:1:1101:10131:20635  II  3958833   \n",
       "\n",
       "                                                 seq strand_x    UMIs  \\\n",
       "0  TATACAATTTTTTGAATCCAAAAACAGTTACAGAATATTTCAAGAA...        -  ACCGAC   \n",
       "1  CTGATTTTACAAGCAACATTACTAACAATTACAGGCCGCGAATATA...        +  CACCAC   \n",
       "2  ATGCACATCACTACCGGGTCTTGGGCAGTGCGATAGCGATGGGATT...        -  TCTGAC   \n",
       "3  ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGA...        +  GCTAAG   \n",
       "4  ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACCAAAAGGA...        +  TGGGTC   \n",
       "\n",
       "                                            UMIs_seq          gene strand_y  \\\n",
       "0  ACCGAC_TATACAATTTTTTGAATCCAAAAACAGTTACAGAATATT...   SPAC26A3.05        +   \n",
       "1  CACCAC_CTGATTTTACAAGCAACATTACTAACAATTACAGGCCGC...     SPBC83.01        -   \n",
       "2  TCTGAC_ATGCACATCACTACCGGGTCTTGGGCAGTGCGATAGCGA...  SPNCRNA.3909        -   \n",
       "3  GCTAAG_ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACC...    SPSNRNA.07        +   \n",
       "4  TGGGTC_ACACGTCAGAAAACACCAGCTGCCCTACATGACACGACC...    SPSNRNA.07        +   \n",
       "\n",
       "   flag_R1  start_read  stop_read RNA_type tail_fromcigar tail_fromGREP  \\\n",
       "0        0     3344283    3344373     mRNA                            T   \n",
       "1       16     1513299    1513389     mRNA                            A   \n",
       "2       16     4268826    4268916    ncRNA                           TT   \n",
       "3        0     3958832    3958922    snRNA                            A   \n",
       "4        0     3958832    3958912    snRNA                            A   \n",
       "\n",
       "  tail_GreporCigar tail_from tail_type_FINAL  tail_len_FINAL  \n",
       "0                      cigar    without_tail               0  \n",
       "1                      cigar    without_tail               0  \n",
       "2                      cigar    without_tail               0  \n",
       "3                      cigar    without_tail               0  \n",
       "4                      cigar    without_tail               0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "R2_R1_merge_sam2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "deltadis32deltaski2_clone1_S93_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lidia/miniconda2/envs/rnatails/lib/python3.7/site-packages/ipykernel_launcher.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/home/lidia/miniconda2/envs/rnatails/lib/python3.7/site-packages/ipykernel_launcher.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "deltadis32deltaski2_clone2_S94_\n",
      "deltalsm1_clone1_S81_\n",
      "deltalsm1_clone2_S82_\n",
      "deltalsm1deltaski2_clone1_S95_\n",
      "deltalsm1deltaski2_clone2_S96_\n",
      "deltaski2_clone1_S83_\n",
      "deltaski2_clone2_S84_\n",
      "Wild_type_clone1_TS_S75_\n",
      "Wild_type_clone2_TS_S76_\n"
     ]
    }
   ],
   "source": [
    "names = [#'deltacid1_clone1_S77_',  'deltacid1_clone2_S78_',  'deltacid1deltadis32_clone1_S85_',  \n",
    "         #'deltacid1deltadis32_clone2_S86_', \n",
    "   # 'deltacid1deltadis32deltalsm1_clone1_S97_',  \n",
    "         #'deltacid1deltadis32deltalsm1_clone2_S98_',  'deltacid1deltadis32deltaski2_clone1_S99_',  \n",
    "        # 'deltacid1deltadis32deltaski2_clone2_S100_',  'deltacid1deltalsm1_clone1_S87_',  \n",
    "         #'deltacid1deltalsm1_clone2_S88_',  'deltacid1deltalsm1deltaski2_clone1_S101_',  \n",
    "        # 'deltacid1deltalsm1deltaski2_clone2_S102_',  'deltacid1deltaski2_clone1_S89_',  \n",
    "        # 'deltacid1deltaski2_clone2_S90_',  'deltadis32_clone1_S79_',  'deltadis32_clone2_S80_',  \n",
    "        # 'deltadis32deltalsm1_clone1_S91_',  'deltadis32deltalsm1_clone2_S92_',  \n",
    "        # 'deltadis32deltalsm1deltaski2_clone1_S103_',\n",
    "    #####'deltadis32deltalsm1deltaski2_clone2_S104_',  \n",
    "         'deltadis32deltaski2_clone1_S93_',  'deltadis32deltaski2_clone2_S94_',  'deltalsm1_clone1_S81_',  \n",
    "         'deltalsm1_clone2_S82_',  'deltalsm1deltaski2_clone1_S95_',  'deltalsm1deltaski2_clone2_S96_', \n",
    "         'deltaski2_clone1_S83_',  'deltaski2_clone2_S84_',  'Wild_type_clone1_TS_S75_', \n",
    "         'Wild_type_clone2_TS_S76_']\n",
    "\n",
    "\n",
    "\n",
    "for name in names:\n",
    "    print(name)\n",
    "    R2_R1_merge_sam2 = pd.read_csv('R2_allsamples/dfs_output/df_{}.csv'.format(name))\n",
    "    \n",
    "    R2_R1_merge_sam2['RNA_type'] = R2_R1_merge_sam2['gene'].apply(lambda x: test_RNA_type(x.lower()))\n",
    "    R2_R1_merge_sam2['tail_fromcigar'] = R2_R1_merge_sam2.apply(lambda kol: take_tail_fromcigar2(kol.strand_y, kol.cigar,kol.seq), axis = 1)\n",
    "    R2_R1_merge_sam2['tail_fromGREP'] = R2_R1_merge_sam2.apply(lambda kol: grep_tail_edit(kol.strand_y,  kol.seq), axis = 1)\n",
    "   # R2_R1_merge_sam2['tail_type_CIGAR'] = R2_R1_merge_sam2.apply(lambda kol: test_tail(kol.strand_y, kol.tail_fromcigar), axis = 1)\n",
    "   # R2_R1_merge_sam2['tail_len_CIGAR'] = R2_R1_merge_sam2['tail_fromcigar'].apply(lambda x: len(x))\n",
    "    #R2_R1_merge_sam2['tail_type_GREP'] = R2_R1_merge_sam2.apply(lambda kol: test_tail(kol.strand_y, kol.tail_fromGREP), axis = 1)\n",
    "   # R2_R1_merge_sam2['tail_len_GREP'] = R2_R1_merge_sam2['tail_fromGREP'].apply(lambda x: len(x))\n",
    "   # R2_R1_merge_sam2 = R2_R1_merge_sam2.drop(columns=['cigar'])\n",
    "    R2_R1_merge_sam2 = R2_R1_merge_sam2.drop_duplicates()\n",
    "    R2_R1_merge_sam2['tail_GreporCigar'] = R2_R1_merge_sam2.apply(lambda kol: tail_fromGREPorCIGAR(kol.cigar, kol.tail_fromGREP,kol.tail_fromcigar), axis = 1)\n",
    "    R2_R1_merge_sam2['tail_from'] = R2_R1_merge_sam2.apply(lambda kol: tail_fromGREPorCIGAR_description(kol.cigar, kol.tail_fromGREP,kol.tail_fromcigar), axis = 1)\n",
    "    R2_R1_merge_sam2['tail_type_FINAL'] = R2_R1_merge_sam2.apply(lambda kol: test_tail(kol.strand_y, kol.tail_GreporCigar), axis = 1)\n",
    "    R2_R1_merge_sam2['tail_len_FINAL'] = R2_R1_merge_sam2['tail_GreporCigar'].apply(lambda x: len(x))\n",
    "    R2_R1_merge_sam2 = R2_R1_merge_sam2.drop(columns=['cigar'])\n",
    "    R2_R1_merge_sam2 = R2_R1_merge_sam2.drop_duplicates()   \n",
    "    R2_R1_merge_sam3 = pd.merge(R2_R1_merge_sam2,reference_newSP,on = ['gene'], how='left')\n",
    "    closest_seq_plus = R2_R1_merge_sam3[R2_R1_merge_sam3['strand_y'] == '+']\n",
    "    closest_seq_minus = R2_R1_merge_sam3[R2_R1_merge_sam3['strand_y'] == '-']\n",
    "\n",
    "    closest_seq_plus['distance'] = -(closest_seq_plus['stop'] - closest_seq_plus['stop_read'])\n",
    "    closest_seq_minus['distance'] = -(closest_seq_minus['start_read'] - closest_seq_minus['start'])\n",
    "\n",
    "    R2_R1_merge_sam4_PAS = pd.concat([closest_seq_minus, closest_seq_plus])\n",
    "    R2_R1_merge_sam4_PAS.to_csv('R2_allsamples/dfs_output/df_{}detailed.csv'.format(name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/lidia/ADATA SD700/tailseq_dec2021_manysamples\r\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/deltadis32deltalsm1deltaski2_clone2_S104_R1_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1deltaski2_clone2_S104_R1_001.fastq.paired.fq.gz\r\n",
      "data/deltadis32deltalsm1deltaski2_clone2_S104_R2_001.fastq.gz\r\n",
      "data/deltadis32deltalsm1deltaski2_clone2_S104_R2_001.fastq.paired.fq\r\n"
     ]
    }
   ],
   "source": [
    "!ls data/*104*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R2_allsamples/dfs_output/df_deltacid1_clone1_S77_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1_clone2_S78_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltadis32_clone1_S85_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltadis32_clone2_S86_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltadis32deltalsm1_clone1_S97_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltadis32deltalsm1_clone2_S98_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltadis32deltaski2_clone1_S99_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltadis32deltaski2_clone2_S100_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltalsm1_clone1_S87_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltalsm1_clone2_S88_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltalsm1deltaski2_clone1_S101_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltalsm1deltaski2_clone2_S102_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltaski2_clone1_S89_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltacid1deltaski2_clone2_S90_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltadis32_clone1_S79_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltadis32_clone2_S80_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltadis32deltalsm1_clone1_S91_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltadis32deltalsm1_clone2_S92_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltadis32deltalsm1deltaski2_clone1_S103_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltadis32deltaski2_clone1_S93_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltadis32deltaski2_clone2_S94_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltalsm1_clone1_S81_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltalsm1_clone2_S82_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltalsm1deltaski2_clone1_S95_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltalsm1deltaski2_clone2_S96_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltaski2_clone1_S83_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_deltaski2_clone2_S84_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_Wild_type_clone1_TS_S75_detailed.csv\r\n",
      "R2_allsamples/dfs_output/df_Wild_type_clone2_TS_S76_detailed.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls R2_allsamples/dfs_output/df_*detailed.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#R2_R1_merge_sam2['RNA_type'] = R2_R1_merge_sam2['gene'].apply(lambda x: test_RNA_type(x.lower()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "R2_R1_merge_sam3_in = R2_R1_merge_sam2.set_index('readID')\n",
    "r2_bed_in = r2_bed.set_index('readID')\n",
    "\n",
    "R2_R1_merge_sam4 = R2_R1_merge_sam3_in.join(r2_bed_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "closest_seq_plus = R2_R1_merge_sam4[R2_R1_merge_sam4['strand_y'] == '+']\n",
    "closest_seq_minus = R2_R1_merge_sam4[R2_R1_merge_sam4['strand_y'] == '-']\n",
    "\n",
    "closest_seq_plus['distance'] = -(closest_seq_plus['stop'] - closest_seq_plus['stop_read'])\n",
    "closest_seq_minus['distance'] = -(closest_seq_minus['start_read'] - closest_seq_minus['start'])\n",
    "\n",
    "R2_R1_merge_sam4_PAS = pd.concat([closest_seq_minus, closest_seq_plus])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
